### PR TITLE
feat: dynamic ground truths

### LINKED ISSUE #13: Chatbot: Dynamic ground truths
https://github.com/ubiquity-os-marketplace/command-ask/blob/e45db813be018fe8772414ca6ec7cbba3b8e55e1/src/adapters/openai/helpers/completions.ts#L40-L42

https://github.com/ubiquity-os-marketplace/command-ask/blob/e45db813be018fe8772414ca6ec7cbba3b8e55e1/src/handlers/ask-llm.ts#L70

I think these should be dynamic, I don't have super clear context on them but from what I can surmise they'd benefit from it.

- Are they restricted to being one-word strings? What's the max length if not? How many truths can we provide/what's the optimal amount?
- Are they supposed to embody the org, the incoming query or the task in relation to the current `@ubqbot` command or the fully built `formattedChat` that includes all context?

I think we have GPT-4o build our array based on the input it's given because:
1. Truths are dynamic and based on either A) fully built/initial incoming user query B) built around the payload issue spec or whatever C) task spec when considering code review
2. We need a quick response imo, a huge spec to truth followed by the spec + the entire diff both to `o1` may be too long to wait for little added benefit.

### PR DESCRIPTION
**Pull request by Keyrxng**
Resolves #13

### CODE REVIEW

**Comment by Keyrxng** [comment]
Marking as ready for ready to get eyes on it and opinions as-is. Will provide QA in a sec, picked from [e6586a4](https://github.com/ubiquity-os-marketplace/command-ask/pull/11/commits/e6586a48570c173a7d797c9ebc6c16c884633727)

**Comment by Keyrxng** [comment]
@0x4007 @sshivaditya2019 @gentlementlegen requesting review

CI can be ignored as it's used in #11 but not here, or I can comment it out or something so it passes CI.

**Comment by 0x4007** [comment]
Why are you adding redundant information in those arrays?

**Comment by Keyrxng** [comment]
> Why are you adding redundant information in those arrays?

I'm not manually adding anything it's GPT that's creating the array contents.

Without the context and add. info I requested here #13, I'm not 100% how to refine and improve inline with @sshivaditya2019' original intention for them.

Right now GPT is consuming the task spec and creating these outputs based on this prompt and settings passed to `openai.createCompletion`

**Comment by Keyrxng** [comment]
the @ubqbot command right now that's in `development` the prompt looks like this regarding ground truths:

`You Must obey the following ground truths: ["typescript" : "github" : "cloudflare worker" : "actions" : "jest" : "supabase": "openai"]`

Which doesn't make a whole lot of sense to me without the additional context. These are more like a classification of the subject areas of the tech stack involved in the query/task/org?

**Comment by 0x4007** [comment]
I think we should not have redundant messages but they should be more substantial than the keywords we have now.

**Comment by Keyrxng** [comment]
In my opinion "Ground Truths" should be considered in relation to the use-case if they are intended to guardrail the model to conform to a specific workflow, which we can consider different applications, i.e chatbot vs code review.

- `@ubqbot`: General Organization Chatbot. It's truths should be like `"The org uses ...these tech stacks only consider these in your response."`, `"This repo uses ...these frameworks/libs"`, etc...
- `pull precheck`: Pull Request Review. It's truths should be based on the task spec (and maybe some of our contribution standards) as that's the source of truth for this application of the model, proving the spec is implemented. It's truths are dynamic, if we had a contributing.md or something in each repo we could ground truth those.

I'll improve the current prompt through #11 and complete this task once it's accepted.

**Comment by Keyrxng** [comment]
Dynamic chatbot ground truths QA:

https://github.com/ubq-testing/ask-plugin/issues/8
https://github.com/ubq-testing/ask-plugin/issues/9


```yml
 [
  {
    role: 'system',
    content: '\n' +
      'Using the input provided, your goal is to produce an array of strings that represent "Ground Truths."\n' +
      'These ground truths are high-level abstractions that encapsulate the tech stack and dependencies of the repository.\n' +
      '  \n' +
      'Each ground truth should:\n' +
      '- Be succinct and easy to understand.\n' +
      '- Use only the information provided in the input.\n' +
      '- Focus on essential requirements, behaviors, or assumptions involved in the repository.\n' +
      '  \n' +
      'Example:\n' +
      'Languages: { TypeScript: 60%, JavaScript: 15%, HTML: 10%, CSS: 5%, ... }\n' +
      'Dependencies: Esbuild, Wrangler, React, Tailwind CSS, ms, React-carousel, React-icons, ...\n' +
      'Dev Dependencies: @types/node, @types/jest, @mswjs, @testing-library/react, @testing-library/jest-dom, @Cypress ...\n' +
      'Ground Truths:\n' +
      '- The repo predominantly uses TypeScript, with JavaScript, HTML, and CSS also present.\n' +
      '- The repo is a React project that uses Tailwind CSS.\n' +
      '- The project is built with Esbuild and deployed with Wrangler, indicating a Cloudflare Workers project.\n' +
      '- The repo tests use Jest, Cypress, mswjs, and React Testing Library.\n' +
      '  \n' +
      'Conditions:\n' +
      'Assume your output builds the foundation for a chatbot to understand the repository when asked an arbitrary query.\n' +
      'Do not list every language or dependency, focus on the most prevalent ones.\n' +
      'Focus on what is essential to understand the repository at a high level.\n' +
      'Brevity is key. Use zero formatting. Do not wrap in quotes, backticks, or other characters.\n' +
      'response === ["some", "array", "of", "strings"]\n' +
      '  \n' +
      'Generate similar ground truths adhering to a maximum of 10.\n' +
      '  \n' +
      'Return a JSON parsable array of strings representing the ground truths, without comment or directive.'
  },
  {
    role: 'user',
    content: '{"dependencies":{"@mswjs/data":"^0.16.2","@octokit/rest":"20.1.1","@octokit/webhooks":"13.2.7","@sinclair/typebox":"0.32.33","@supabase/supabase-js":"^2.45.4","@ubiquity-dao/ubiquibot-logger":"^1.3.0","dotenv":"^16.4.5","openai":"^4.63.0","typebox-validators":"0.3.5","voyageai":"^0.0.1-5"},"devDependencies":{"@actions/core":"^1.11.1","@actions/github":"^6.0.0","@commitlint/cli":"19.3.0","@commitlint/config-conventional":"19.2.2","@cspell/dict-node":"5.0.1","@cspell/dict-software-terms":"3.4.6","@cspell/dict-typescript":"3.1.5","@eslint/js":"9.5.0","@jest/globals":"29.7.0","@types/jest":"^29.5.12","@types/node":"20.14.5","cspell":"8.9.0","eslint":"9.5.0","eslint-config-prettier":"9.1.0","eslint-plugin-check-file":"2.8.0","eslint-plugin-prettier":"5.1.3","eslint-plugin-sonarjs":"1.0.3","husky":"9.0.11","jest":"29.7.0","jest-junit":"16.0.0","jest-md-dashboard":"0.8.0","knip":"5.21.2","lint-staged":"15.2.7","npm-run-all":"4.1.5","prettier":"3.3.2","ts-jest":"29.1.5","tsx":"4.15.6","typescript":"5.4.5","typescript-eslint":"7.13.1","wrangler":"^3.81.0"},"languages":[["TypeScript",0.9235672829913418],["PLpgSQL",0.03861807956191261],["JavaScript",0.03622889642996839],["Shell",0.00158574101677714]]}'
  }
]
languages:  [                                                                                                                                                                               
  [ 'TypeScript', 0.9235672829913418 ],                                                                                                                                                     
  [ 'PLpgSQL', 0.03861807956191261 ],                                                                                                                                                       
  [ 'JavaScript', 0.03622889642996839 ],
  [ 'Shell', 0.00158574101677714 ]
]
Ground Truths:  [
  'The repository is primarily written in TypeScript, with some PLpgSQL and JavaScript code.',
  'The project uses Supabase for backend services.',
  'Integration with GitHub APIs is handled via Octokit.',
  "The application leverages OpenAI's API for AI functionalities.",
  'Jest is used as the testing framework, configured for TypeScript.',
  'ESLint and Prettier are employed for code linting and formatting.',
  'GitHub Actions manage the CI/CD workflows.',
  'Husky and lint-staged are set up for pre-commit hooks.',
  'The project is deployed using Wrangler, indicating deployment to Cloudflare Workers.',
  'Commit messages are enforced using Commitlint with conventional commit standards.'
]```

```yml
 [
  {
    role: 'system',
    content: '\n' +
      'Using the input provided, your goal is to produce an array of strings that represent "Ground Truths."\n' +
      'These ground truths are high-level abstractions that encapsulate the tech stack and dependencies of the repository.\n' +
      '  \n' +
      'Each ground truth should:\n' +
      '- Be succinct and easy to understand.\n' +
      '- Use only the information provided in the input.\n' +
      '- Focus on essential requirements, behaviors, or assumptions involved in the repository.\n' +
      '  \n' +
      'Example:\n' +
      'Languages: { TypeScript: 60%, JavaScript: 15%, HTML: 10%, CSS: 5%, ... }\n' +
      'Dependencies: Esbuild, Wrangler, React, Tailwind CSS, ms, React-carousel, React-icons, ...\n' +
      'Dev Dependencies: @types/node, @types/jest, @mswjs, @testing-library/react, @testing-library/jest-dom, @Cypress ...\n' +
      'Ground Truths:\n' +
      '- The repo predominantly uses TypeScript, with JavaScript, HTML, and CSS also present.\n' +
      '- The repo is a React project that uses Tailwind CSS.\n' +
      '- The project is built with Esbuild and deployed with Wrangler, indicating a Cloudflare Workers project.\n' +
      '- The repo tests use Jest, Cypress, mswjs, and React Testing Library.\n' +
      '  \n' +
      'Conditions:\n' +
      'Assume your output builds the foundation for a chatbot to understand the repository when asked an arbitrary query.\n' +
      'Do not list every language or dependency, focus on the most prevalent ones.\n' +
      'Focus on what is essential to understand the repository at a high level.\n' +
      'Brevity is key. Use zero formatting. Do not wrap in quotes, backticks, or other characters.\n' +
      'response === ["some", "array", "of", "strings"]\n' +
      '  \n' +
      'Generate similar ground truths adhering to a maximum of 10.\n' +
      '  \n' +
      'Return a JSON parsable array of strings representing the ground truths, without comment or directive.'
  },
  {
    role: 'user',
    content: '{"dependencies":{"@mswjs/data":"^0.16.2","@octokit/rest":"20.1.1","@octokit/webhooks":"13.2.7","@sinclair/typebox":"0.32.33","@supabase/supabase-js":"^2.45.4","@ubiquity-dao/ubiquibot-logger":"^1.3.0","dotenv":"^16.4.5","openai":"^4.63.0","typebox-validators":"0.3.5","voyageai":"^0.0.1-5"},"devDependencies":{"@actions/core":"^1.11.1","@actions/github":"^6.0.0","@commitlint/cli":"19.3.0","@commitlint/config-conventional":"19.2.2","@cspell/dict-node":"5.0.1","@cspell/dict-software-terms":"3.4.6","@cspell/dict-typescript":"3.1.5","@eslint/js":"9.5.0","@jest/globals":"29.7.0","@types/jest":"^29.5.12","@types/node":"20.14.5","cspell":"8.9.0","eslint":"9.5.0","eslint-config-prettier":"9.1.0","eslint-plugin-check-file":"2.8.0","eslint-plugin-prettier":"5.1.3","eslint-plugin-sonarjs":"1.0.3","husky":"9.0.11","jest":"29.7.0","jest-junit":"16.0.0","jest-md-dashboard":"0.8.0","knip":"5.21.2","lint-staged":"15.2.7","npm-run-all":"4.1.5","prettier":"3.3.2","ts-jest":"29.1.5","tsx":"4.15.6","typescript":"5.4.5","typescript-eslint":"7.13.1","wrangler":"^3.81.0"},"languages":[["TypeScript",0.9235672829913418],["PLpgSQL",0.03861807956191261],["JavaScript",0.03622889642996839],["Shell",0.00158574101677714]]}'
  }
]
languages:  [                                                                                                                                                                               
  [ 'TypeScript', 0.9235672829913418 ],                                                                                                                                                     
  [ 'PLpgSQL', 0.03861807956191261 ],                                                                                                                                                       
  [ 'JavaScript', 0.03622889642996839 ],
  [ 'Shell', 0.00158574101677714 ]
]
Ground Truths:  [
  'The repository is primarily written in TypeScript with minor use of JavaScript and PLpgSQL.',
  'It integrates with Supabase for backend services.',
  'The project leverages OpenAI for AI functionalities.',
  'Environment variables are managed using dotenv.',
  'Deployment is handled with Wrangler, indicating Cloudflare Workers usage.',
  'The development setup includes Jest for testing and ESLint for linting.',
  'GitHub Actions are employed for continuous integration and deployment workflows.',
  'Commit messages are standardized using Commitlint and enforced with Husky hooks.',
  'The project uses @octokit libraries for GitHub API interactions and webhooks.',
  'TypeScript is utilized with typebox for schema validation and type safety.'
]
```

File: src/handlers/ground-truths/prompts.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import { GroundTruthsSystemMessageTemplate, ModelApplications } from "../../types/llm";
+
+const CODE_REVIEW_GROUND_TRUTHS_SYSTEM_MESSAGE = {
```
Comment (Keyrxng) [inline src/handlers/ground-truths/prompts.ts:3]: This can be ignored as it's only relevant to #11 so this is all the work on it I'll do here. 

lmk if it's better to remove all #11 related logic and I will.

File: src/types/llm.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  taskSpecification: string;
+};
+
+export type AppParamsHelper<TApp extends ModelApplications> = TApp extends "code-review"
+  ? CodeReviewAppParams
+  : TApp extends "chat-bot"
+    ? ChatBotAppParams
+    : never;
+
+export type CompletionsModelHelper<TApp extends ModelApplications> = TApp extends "code-review" ? "gpt-4o" : TApp extends "chat-bot" ? "o1-mini" : never;
```
Comment (Keyrxng) [inline src/types/llm.ts:21]: I implemented this thinking it'd be too slow but I think `o1-mini` is fast enough actually. If this is not helpful I'll remove it. 

1. This enforces the correct usage at the type level.
2. Hardcoded but might be able to make it dynamic if required.

**Comment by Keyrxng** [comment]
@gentlementlegen @0x4007 @sshivaditya2019 @rndquu 

Why can't we request reviews in this org lmao? Anyway this is ready for review team, thanks.

File: src/handlers/ground-truths/chat-bot.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  });
+
+  if ("content" in packageJson) {
+    return extractDependencies(JSON.parse(Buffer.from(packageJson.content, "base64").toString()));
+  } else {
+    throw logger.error(`No package.json found in ${owner}/${repo}`);
+  }
+}
+
+export function extractDependencies(packageJson: Record<string, Record<string, string>>) {
```
Comment (Keyrxng) [inline src/handlers/ground-truths/chat-bot.ts:28]: Broke this into a fn to make it easier to see what we are pulling from the `package.json`.

I'm thinking it might have value to also include at least some of these, lmk which ones and if you have a prompt structure in mind also lmk.


```json
{
"name": "@ubiquity-os/command-ask",
  "version": "1.0.0",
  "description": "A highly context aware organization integrated chatbot",
  "author": "Ubiquity DAO",
  "license": "MIT",
  "main": "src/worker.ts",
  "engines": {
    "node": ">=20.10.0"
  },
"keywords": [
    "typescript",
    "template",
    "dao",
    "ubiquity",
    "open-source"
  ],
```

File: src/handlers/ground-truths/chat-bot.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        owner: { login: owner },
+        name: repo,
+      },
+    },
+  } = context;
+
+  const { data: packageJson } = await octokit.repos.getContent({
+    owner,
+    repo,
+    path: "package.json",
```
[the remark this answers is not in the document]
Comment (Keyrxng) [reply src/handlers/ground-truths/chat-bot.ts:18]: We should either

A. Just hardcode a list of these sorts of files.
B. New config item we read them from with a fallback to the hardcoded list.

These files would require individual processing I'd imagine so either

C. We don't extract anything and we just feed the whole file to the model. Unclear to me what most other files look like.
D. We implement manual handling for each hardcoded path.

---

Either way, this should be a separate task as it'll involve quite a bit of work.

File: src/handlers/ground-truths/prompts.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    "Languages: { TypeScript: 60%, JavaScript: 15%, HTML: 10%, CSS: 5%, ... }",
+    "Dependencies: Esbuild, Wrangler, React, Tailwind CSS, ms, React-carousel, React-icons, ...",
+    "Dev Dependencies: @types/node, @types/jest, @mswjs, @testing-library/react, @testing-library/jest-dom, @Cypress ...",
+    "Ground Truths:",
+    "- The repo predominantly uses TypeScript, with JavaScript, HTML, and CSS also present.",
+    "- The repo is a React project that uses Tailwind CSS.",
+    "- The project is built with Esbuild and deployed with Wrangler, indicating a Cloudflare Workers project.",
+    "- The repo tests use Jest, Cypress, mswjs, and React Testing Library.",
+  ],
+  conditions: [
```
[the remark this answers is not in the document]
Comment (Keyrxng) [reply src/handlers/ground-truths/prompts.ts:45]: Can you show me what you mean and convert one or two of those lines into what you mean exactly please bud? I don't quite understand as the condition is the list item

[the remark this answers is not in the document]
Comment (Keyrxng) [reply src/handlers/ground-truths/prompts.ts:45]:
```
1. ASSUME your output BUILDS the FOUNDATION for a chatbot to UNDERSTAND the REPOSITORY when asked an arbitrary QUERY.
2. DO NOT LIST every LANGUAGE or DEPENDENCY; FOCUS on the most PREVALENT ones.
3. FOCUS on what is ESSENTIAL to UNDERSTAND the REPOSITORY at a HIGH LEVEL.
4. BREVITY is KEY. USE NO FORMATTING — DO NOT WRAP in quotes, BACKTICKS, or other CHARACTERS.
5. EXPECTED FORMAT: response === ["some", "array", "of", "strings"].
```

Will that do? 

I'm aware capitalizing has it's benefits but haven't seen it used like this before, would be cool if you could link to the systems you are talking about.

I wouldn't have done this personally but if you think it'll be an improvement I will implement your approved conditions ^.

File: src/types/llm.ts (hunk 2 above)
Comment (Keyrxng) [reply src/types/llm.ts:21]: The suggestion is to use claude when parsing code as was recommended to me by @0x4007 however `o1-mini` will probably be used instead as it's proven quick enough so far in my testing.

In #11 we are passing the entire PR diff + the task spec which has had truths generated from it also that exist within the system message. Are you guys suggesting that I pass the spec + diff to claude instead of GPT which by my understanding is the more advanced model?

I ask because it's not just parsing the code, there is a very complex degree of reasoning etc required to be able to tie the changes to the spec effectively. Is Claude sufficient for such large tasks?

Comment (Keyrxng) [reply src/types/llm.ts:21]: What should be fed to Claude is my base question: just code? or code and any other context and requests that we'd ask of GPT?

File: src/handlers/ground-truths/prompts.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Keyrxng) [reply src/handlers/ground-truths/prompts.ts:45]: I've used/seen it used sparingly throughout a prompt to enforce a nuance in formatting like `NEVER wrap in backticks` and that sort of thing or to really drive home one or two points but I do agree it produces a lot of value when used well.

Since this is a "test in production" sort of plugin, what I'll do is embed these ground truths into the final llm response and for when/if they are revisited later we'll have a good dataset to be able to refine the prompt with. (a set of truths is created with every query)

File: src/types/llm.ts (hunk 2 above)
Comment (Keyrxng) [reply src/types/llm.ts:21]: So to clarify, for #11, the context window will contain:

- raw task spec
- ground truths sourced from the task spec
- code review prompt/instructions
- raw PR diff

These should all be fed to Claude 3.5 instead of GPT, and Claude will be performing our code reviews (comparing spec to changes made, not writing code) start-to-finish (minus ground-truth sourcing)?


> I think it would be best if any information related to a coding task is passed to Claude, while we default to the other OpenAI model for everything else.

Your usage of task here confuses me sorry. All of our github issues are coding tasks. Do you mean any queries that involve raw code should be fed to Claude instead of GPT?

Comment (Keyrxng) [reply src/types/llm.ts:21]: Okay this will need another task opened to alter the `@ubqbot` logic as all context goes to `o1-mini` including pull request diffs currently.

Would you like to make a task for it?

File: src/handlers/ground-truths/prompts.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (0x4007) [reply src/handlers/ground-truths/prompts.ts:45]: OpenAI system prompts don't look like this so I'm inclined to believe that's not accurate.

[the remark this answers is not in the document]
Comment (Keyrxng) [reply src/handlers/ground-truths/prompts.ts:45]: As am I but would you want me to still embed the truths that get sourced at least initially?

Also, we used to escape the OpenAI system message. We should probably bring that back.

File: src/types/llm.ts (hunk 2 above)
Comment (0x4007) [reply src/types/llm.ts:21]: In my deep experience, when I am building off of and integrating/editing existing code I always use Claude. It's careful and considerate of existing code. But if I'm one-shotting a brand new feature that doesn't import/edit existing code I'll use o1-preview. It's really good at being its own leader where Claude is much better as an "assistant" 

If I need to pass in some code context but it's not directly being used/imported in the newly generated code, I use o1-mini because it has a larger usable context window (both preview and mini have ~65k tokens per request but preview eats up most of them for internal reasoning steps)

In conclusion, when dealing with specific details of code, vs the "general idea of it" always use Claude. It's far more precise. 

Hope that helps.

Comment (Keyrxng) [reply src/types/llm.ts:21]: `@ubqbot` because it's pulling in context from multiple task specs, multiple convos and PRs (raw code) I think GPT is also more appropriate but I've never used Claude.

---

By my understanding of the feature introduced in #11 we need it to be able to compartmentalize the spec and the code and make very accurate deductions regarding it's implementation, like your description of `o1` being it's own leader.

nuances with code review:
- specs may be lacking and/or hard to extrapolate from, requires in-prod testing
- we need it to enforce the spec but not to the letter as it'll bounce PRs for too many false flags (as I've witnessed building the last review logic)

File: src/handlers/ground-truths/prompts.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (0x4007) [reply src/handlers/ground-truths/prompts.ts:45]: We can try ground truths I was referring to capitalized words for emphasis.

[the remark this answers is not in the document]
Comment (Keyrxng) [reply src/handlers/ground-truths/prompts.ts:45]: I was referring to that also my question was referring to meaning using the SDK comment metadata feature

**Comment by Keyrxng** [comment]
https://github.com/ubq-testing/ask-plugin/issues/11#issuecomment-2438917655

with every usage of `@ubqbot` we get (I've wrapped in backticks so you all can see)

```
<!-- Ubiquity - LLM Ground Truths and Token Usage - runPlugin - undefined
{
  "metadata": {
    "groundTruths": [
      "The repository is primarily written in TypeScript with some PLpgSQL and JavaScript.",
      "Supabase is used for backend services and database management.",
      "GitHub Actions are integrated for continuous integration and deployment workflows.",
      "Jest is utilized for testing the codebase.",
      "ESLint and Prettier are employed for code linting and formatting.",
      "The project leverages OpenAI APIs for its functionalities.",
      "Wrangler indicates that the project is deployed on Cloudflare Workers.",
      "TypeBox is used for type definitions and schema validations.",
      "Husky and lint-staged manage Git hooks and enforce code quality.",
      "Environment variables are handled using dotenv."
    ],
    "tokenUsage": {
      "input": 923,
      "output": 46,
      "total": 969
    }
  },
  "caller": "runPlugin"
}
-->
```

File: src/types/llm.ts (hunk 2 above)
Comment (Keyrxng) [reply src/types/llm.ts:21]: Resolving this thread as this requires it's own task as there is a lot of discussion and potential work involved in categorizing queries on the fly and determining if they should be fed to Claude or GPT. It should not hold back this PR.

**Comment by 0x4007** [comment]
@sshivaditya2019 you should review and decide when this pull is ready. I encourage QA for changes to prove they work and ideally you should also test as a reviewer

**Comment by Keyrxng** [comment]
lmk if there is anything holding back this PR and I'll push it forward
